azure-auth-helper/src/main/java/com/microsoft/azure/auth/LocalAuthServer.java

```diff
@@ -0,0 +1,149 @@
+import java.io.PrintWriter;
+import java.net.URI;
+import java.util.concurrent.Semaphore;
+import java.util.concurrent.TimeUnit;
+
+public class LocalAuthServer {
+    private static transient boolean inited = false;
+    private static Object initLock = new Object();
+    private static String loginSuccessBodyTemplate;
+    private static String loginErrorBodyTemplate;
+    private final Semaphore semaphore = new Semaphore(0);
+    private final Server jettyServer;
+
+    private String code;
+    private String error;
+    private String errorDescription;
+
+    public LocalAuthServer() {
+        jettyServer = new Server();
+        final ServerConnector connector = new ServerConnector(jettyServer);
+        connector.setHost("localhost");
+        jettyServer.setConnectors(new Connector[]{ connector });
+        jettyServer.setHandler(new AbstractHandler() {
+            @Override
+            public void handle(String target, Request baseRequest, HttpServletRequest request, HttpServletResponse response)
+                    throws IOException, ServletException {
+                initHtmlTemplate();
+                try {
+                    ((Request) request).setHandled(true);
+                    code = request.getParameter(Constants.CODE);
+                    // handle the error response described at
+                    // https://docs.microsoft.com/en-us/azure/active-directory/develop/v1-protocols-oauth-code#error-response
+                    errorDescription = request.getParameter(Constants.ERROR_DESCRIPTION);
+                    error = request.getParameter(Constants.ERROR);
+
+                    final boolean isSuccess = StringUtils.isEmpty(error) && StringUtils.isNotEmpty(code);
+                    response.setStatus(HttpServletResponse.SC_OK);
+                    response.setContentType(Constants.CONTENT_TYPE_TEXT_HTML);
+                    try (final PrintWriter writer = response.getWriter()) {
+                        writer.write(isSuccess ? loginSuccessBodyTemplate : String.format(loginErrorBodyTemplate, error, errorDescription));
+                        writer.flush();
+                    }
+                    response.flushBuffer();
+                } finally {
+                    semaphore.release();
+                }
+            }
+
+        });
+    }
+
+    public URI getURI() {
+        return jettyServer.getURI();
+    }
+
+    public void start() throws IOException {
+        if (jettyServer.isStarted()) {
+            return;
+        }
+        try {
+            jettyServer.start();
+        } catch (Exception e) { // server.start will throw Exception
+            if (e instanceof RuntimeException) {
+                // avoid unnecessary exception convert
+                throw (RuntimeException) e;
+            }
+
+            throw new IOException(e);
+        }
+
+    }
+
+    public void stop() throws IOException {
+        semaphore.release();
+        try {
+            jettyServer.stop();
```

Comment:
There are multiple callers to the stop function. Is it safe to call jettyServer.stop() again and again?

Reply:
safe for multiple calls: 
```java
public final void stop() throws Exception
    {
        synchronized (_lock)
        {
            try
            {
                if (_state == __STOPPING || _state == __STOPPED)
                    return;
                setStopping();
                doStop();
                setStopped();
            }
            catch (Throwable e)
            {
                setFailed(e);
                throw e;
            }
        }
    }
```